### PR TITLE
Homework1

### PR DESCRIPTION
**Pull request by ArtemTsukanov**
* Выполнено EDA в ноутбуке `notebooks/experiments.ipynb`, а также сгенериван отчет о проведенном анализе в `reports/experiments.pdf`.
* Написан пайплайн для тренировки модели в `heart_classification/train_pipeline.py` и тесты к нему, включая прогон синтетических данных.
* Написан метод для получения предсказаний в `heart_classification/predict.py` и соответствующие тесты.
* Для запуска пайплайна и предсказаний используется Hydra с конфигами `configs/train_config.yml` и `configs/predict_config.yml`, соответственно.
* Так как данные достаточно простые, был написан кастовый трансформер, принимающие категориальные и численные фичи, в `heart_classification/entities/feature_transformer.py`.
* Настроен прогон линтера и всех тестов на основе `GitHub Actions`.

Задачи выполнены полностью по списку, поэтому, по моему мнению, стоит оценить работу в 38 баллов (за вычетом возможных недочетов).

### CODE REVIEW

File: ml_project/heart_classification/entities/feature_transformer.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        self.categorical_features = categorical_features
+        self.numerical_features = numerical_features
+        self.encoder = OneHotEncoder(sparse=False, handle_unknown='ignore')
+
+    def fit(self, X: pd.DataFrame):
+        self.encoder.fit(X[self.categorical_features])
+        return self
+
+    def transform(self, X: pd.DataFrame) -> pd.DataFrame:
+        df_num = X[self.numerical_features].copy()
```
Comment (nickdndev) [inline ml_project/heart_classification/entities/feature_transformer.py:29]: необходимо добавить проверку , что fit был вызван + pytest
https://github.com/scikit-learn/scikit-learn/blob/15a949460/sklearn/preprocessing/_data.py#L432

File: ml_project/heart_classification/models/fit_predict_model.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+SklearnClassifier = Union[LogisticRegression, SVC]
+
+
+def train_model(
+    features: pd.DataFrame,
+    target: pd.Series,
+    train_params: TrainingParams
+) -> SklearnClassifier:
+    if train_params.model_type == 'LogisticRegression':
+        model = LogisticRegression(
```
Comment (nickdndev) [inline ml_project/heart_classification/models/fit_predict_model.py]: не стоит харкодить модели в коде.
стоит использовать instantiate_objects hydra и передавать параметры: таргет класс модели и его параметры 
https://hydra.cc/docs/patterns/instantiate_objects/overview/

**Review by nickdndev — COMMENTED** [review]
1. не нашел скрипта , который сгенерил отчет . ("закоммитьте и скрипт и отчет (за это + 1 балл)")
2. необходимо использовать instantiate_objects вместо хардкода моделей в коде 
3. зачем делать gitkeep ? если юзер будет использовать дефолтные директории , то должны создаваться директории т.к. сейчас директории просто "подвисли"
4. если это init PR, то необходимо удалить код в master т.к. сейчас идет diff

Остальное все отлично , согласно заявленной самооценке

File: ml_project/heart_classification/predict.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import hydra
+from omegaconf import OmegaConf
+import yaml
+
+from heart_classification.entities import PredictionParams
+from heart_classification.data import read_data
+from heart_classification.features import build_transformer, make_features
+from heart_classification.models import predict_model
+
+LOGGING_CONFIG_PATH = '../configs/logging_config.yaml'
```
Comment (nickdndev) [inline ml_project/heart_classification/predict.py]: а где используется ?

File: ml_project/heart_classification/train_pipeline.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    make_features,
+    extract_target,
+)
+from heart_classification.models import (
+    train_model,
+    predict_model,
+    evaluate_model,
+)
+
+LOGGING_CONFIG_PATH = '../configs/logging_config.yaml'
```
Comment (nickdndev) [inline ml_project/heart_classification/train_pipeline.py]: а где используется ?

**Comment by ArtemTsukanov** [comment]
.gitkeep для сохранения структуры проекта согласно README,  init-коммит отправлен в мастер исключительно для удобства проверяющих :)

File: ml_project/heart_classification/entities/feature_transformer.py (hunk 1 above)
Comment (ArtemTsukanov) [reply ml_project/heart_classification/entities/feature_transformer.py:29]: поправил

File: ml_project/heart_classification/models/fit_predict_model.py (hunk 2 above)
Comment (ArtemTsukanov) [reply ml_project/heart_classification/models/fit_predict_model.py]: тут тоже, спасибо за совет!

File: ml_project/heart_classification/predict.py (hunk 3 above)
Comment (ArtemTsukanov) [reply ml_project/heart_classification/predict.py]: нигде :)
убрал строчку, было перенесено в конфиг

File: ml_project/heart_classification/train_pipeline.py (hunk 4 above)
Comment (ArtemTsukanov) [reply ml_project/heart_classification/train_pipeline.py]: аналогично

**Comment by nickdndev** [comment]
> .gitkeep для сохранения структуры проекта согласно README, init-коммит отправлен в мастер исключительно для удобства проверяющих :)

1. В продакшине на любом проекте запрещено заливать в мастер до PR т.к. это усложняет ревью 
пример:
https://github.com/made-ml-in-prod-2021/art.tsukanov/blob/main/ml_project/src/data/make_dataset.py в мастере есть , но в PR ничего нет т.к. PR показыкает diff c master

2. Если клиент апи будет использовать свои пути  , то дефолтные директории просто весят в воздухе с gitkeep . Странно это смотрится, но интересно узнать требования от преподавателя о проекте

**Review by Danil328 — APPROVED** [review]
Оценку подтверждаю, 38 баллов

**Comment by Danil328** [comment]
Поставь только labels hw1
image
